Crossfade feature, image clips, auto interval inference

Sv3n opened the pull request:
* Crossfade feature for moviepy backend. Set `crossfade_duration` on the clip that should be faded into, e.g. the second, clip in a list. See samples/round13.json for an example.
* Moviepy backend now supports image clips.
* Automatically complete missing json duration and interval fields where possible.
* Title slides / videos are now also cached.
* Example question now also has spacer.
* Improved moviepy text style.
* Changed travis config to test on xenial: mostly because trusty doesn't have an ffmpeg package, which makes installing it for testing needlessly complicated.
* Fixed bug where adding text on top of a composite video would remove the fade effect from the composite.

This merge request absorbed a little more changes than I originally intended. I hope it is still clear.

Sv3n committed:
Added initial support for crossfading.

Sv3n committed:
Aligned function signatures.

Sv3n committed:
Separate fade-in and fade-out to allow cross-fade.

Sv3n committed:
Silent clip support for moviepy, and restored audio extraction feature.

Sv3n committed:
Add option draw additional answer labels, crossfade test

Sv3n committed:
Support basic comments in json files

Commented line has to start with // to be interpreted as comment.
Optional whitespace prefix is ignored.

Sv3n committed:
Support image as source type

Sv3n committed:
Install ffmpeg explicitly in attempt to fix CI

Sv3n committed:
Use apt-get

Sv3n committed:
Use apt-update

Sv3n committed:
Use xenial

Sv3n committed:
Ignore false warning from pylint, force FPS on clips

Sv3n committed:
Remove useless-pass statements (pylint warning)

Sv3n committed:
Remove unused arguments, increase spacer duration.

Sv3n committed:
Use the packages addon in travis yml

Without update to perhaps save time.

Sv3n committed:
Also include spacer between example question and answer

Sv3n committed:
Assume audio and video interval are the same if one is not specified

Also: auto infer duration of image sources based on their use as clip.

Sv3n committed:
Fix effects on composite video's getting deleted.

Using the custom concatenation did not preserve effects applied to the
composite.
Lowered opacity of text background slightly.
Added stroke around text, changed font to something more fun than Arial.

Sv3n committed:
Do not re-generate theme video's if use_cached_video_files is set

Sv3n committed:
Workaround issue where video's gain duration through rounding

The rounding causes ffmpeg to misreport the actual data content
length of the intermediate video. This causes audio glitches near the
end of the video where the data runs out before the assumed duration.

The workaround is not perfect, occasional glitches remain, but are
less prevalent. Still looking for a better fix.

Sv3n committed:
Fix pylint issues

Sv3n committed:
Small update to readme

Sv3n committed:
Fix mardown syntax in readme

Sv3n committed:
Filter-out question w/o audio spec

Sv3n committed:
Use standard mardown section headings

Sv3n committed:
Modify example to demonstrate automatic interval determination

Sv3n committed:
Pylint fix

CNugteren commented:
Looks good, great additions! :-)

I've checked out your branch and ran your sample it on Linux with ffmpeg, and everything works out fine. I just don't see the `Original artist: Commodores` appearing anywhere in the video. Also this way it won't show up in the questions/answers sheet, or is that on purpose?

Sv3n commented:
Thanks! :)

To get the `answer_label_events` to show up, like the `Original artist: Commodores` you mention, a backend needs to implement `overlay_fading_text(self, text: str, interval: Tuple[float, float])`. The missing part on the ffmpeg side is the support for an explicit effect interval smaller than the size of the full clip. I could not find an existing function in the `FFMpeg` backend that does something like that, so there wasn't a straightforward way for me to implement it easily. I was hoping you would be able to figure it out later ;).

As a fallback, I could try to merge these two if-statements in `filter_stream_video()` into one, i.e. add an optional "interval"  argument to `draw_text` (which does nothing in the ffmpeg backend), and a "height" to `overlay_fading_text`. At that point there is no need to have them as separate functions anymore, and text will least be visible in all output videos:

```python
    if kind == "answer" and answer_label_events is not None:
        for event in answer_label_events:
            evt_interval = event["interval"]
            interval_sec = pypopquiz.io.get_interval_in_fractional_s(evt_interval)
            offset_interval_sec = (interval_sec[0] - interval[0], interval_sec[1] - interval[0])
            ppq.io.log('overlay_fading_text: {}'.format(event["answer"]))
            stream.overlay_fading_text(event["answer"], interval=offset_interval_sec)
    if kind == "answer":
        # (up to the) first two answers are joined together with " - " and shown at the top
        answer_text = " - ".join(answer_texts[:2])
        stream.draw_text_in_box(answer_text, get_interval_length(interval), move=False, top=True,
                                delay_in_sec=delay_answer_text_s)
        # Remainder is shown in the center of the video
        for text_id, answer_text in enumerate(answer_texts[2:]):
            stream.draw_text(answer_text, 0.5 - 0.1 * len(answer_texts[2:]) + 0.2 * text_id)
```

What do you think?

About the answers sheets: I haven't really looked at it yet, but I am aware of the issue. Sample round 7 (beeps) has a similar problem. Having the ability to point at an additional list of answer strings (in this case, in 'answer_label_events') in the json file would solve both at the same time.

Sv3n committed:
Draw answer labels using draw_text instead of overlay_fading_text

Now they also appear using the ffmpeg backend, and there is less code duplication.

CNugteren commented:
OK, sounds good. I guess you already did it, right (judging by the code)?

I've run your 3 latest samples and they all work fine. But in some cases with multi-video answers, you can also nicely give the answers (e.g. original artists) as a second item in the `answers` list, or does that not work in that case?

Sv3n commented:
> 
> 
> But in some cases with multi-video answers, you can also nicely give the answers (e.g. original artists) as a second item in the `answers` list, or does that not work in that case?

Sometimes it works, sometimes it doesn't. Thing get complicated when:

* The number of answers per question is not always the same per round: the `"questioned"` variable in the JSON file needs to be somehow made aware of that in that case.
* When you want the answers to disappear after a while, or to appear consecutively, one at a time. `round07.json` (beeps) is an example of that case. Adding an interval attribute to the 'answers' object could help a little bit, but I like the idea of specifying the time at which an answer appears based on the clip it is drawn over.  You can then time your answers by simply looking at the youtube timecode you want them to appear at. Moving the answer text and interval specification away from the clip specification and into the separate 'answers' block removes the immediate link to the clip, and we would have to connect them again through some other means.

But I agree, this is not ideal yet. Working on the answer sheets problem might bring some inspiration, although my first new thing to work on will either be audio normalization in moviepy, or speeding things up (moviepy backend is still 50x slower than ffmpeg (imho the output just looks much better ;))
